### PR TITLE
Add audit_rules_unsuccessful_file_modification_detailed remediation scripts

### PR DESCRIPTION
**Pull request by redhatrises**

### CODE REVIEW

**Comment by jan-cerny** [comment]
Why is this labeled as a "blocker"? AFAIK pull requests with "blocker" label are impediments to upstream release like failure to build content, or content built is out of standard syntax. Your patch only adds new remediation scripts. I think it isn't a blocker then. What am I missing here?

File: shared/templates/create_audit_rules_unsuccessful_file_modification_detailed.py [hunk 1]
```diff
@@ -37,6 +37,53 @@ def generate(self, target, args):
                 },
                 "./oval/audit_rules_unsuccessful_file_modification_{0}_rule_order.xml", syscall
             )
+
+        elif target == "bash":
+            self.file_from_template(
+                "./template_BASH_audit_rules_unsuccessful_file_modification_o_creat",
```
Comment (jan-cerny) [inline shared/templates/create_audit_rules_unsuccessful_file_modification_detailed.py]: Please also remove `remediation = none` form the affected test scenarios.

File: shared/templates/template_ANSIBLE_audit_rules_unsuccessful_file_modification_o_trunc_write [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    - "-a always,exit -F arch=b32 -S {{{ SYSCALL }}} -F a2&01003 -F exit=-EACCES -F auid>={{{ auid }}} -F auid!=unset -F key=unsuccesful-create"
+    - "-a always,exit -F arch=b32 -S {{{ SYSCALL }}} -F a2&01003 -F exit=-EPERM -F auid>={{{ auid }}} -F auid!=unset -F key=unsuccesful-create"
+
+- name: Inserts/replaces the {{{ SYSCALL }}} rule in audit.rules when on x86_64
+  lineinfile:
+    line: "{{ item }}"
+    state: present
+    dest: /etc/audit/audit.rules
+    create: yes
+  wih_items:
```
Comment (jan-cerny) [inline shared/templates/template_ANSIBLE_audit_rules_unsuccessful_file_modification_o_trunc_write]: It should be `with_items`.

File: shared/templates/create_audit_rules_unsuccessful_file_modification_detailed.py (hunk 1 above)
Comment (jan-cerny) [inline shared/templates/create_audit_rules_unsuccessful_file_modification_detailed.py]: Please remove `#remediation = none` from the test scenarios for these rules.

File: shared/templates/create_audit_rules_unsuccessful_file_modification_detailed.py [hunk 3]
```diff
@@ -37,6 +37,53 @@ def generate(self, target, args):
...
+            )
+            self.file_from_template(
+                "./template_BASH_audit_rules_unsuccessful_file_modification_o_trunc_write",
+                {
+                    "SYSCALL":  syscall
+                },
+                "./bash/audit_rules_unsuccessful_file_modification_{0}_o_trunc_write.sh", syscall
+            )
+            #self.file_from_template(
+            #    "./template_BASH_audit_rules_unsuccessful_file_modification_rule_order",
```
Comment (jan-cerny) [inline shared/templates/create_audit_rules_unsuccessful_file_modification_detailed.py]: We will need these as well, bu let's do that in a separate PR.

Comment (redhatrises) [reply shared/templates/create_audit_rules_unsuccessful_file_modification_detailed.py]: @jan-cerny exactly.

File: shared/templates/template_ANSIBLE_audit_rules_unsuccessful_file_modification_o_trunc_write (hunk 2 above)
Comment (redhatrises) [reply shared/templates/template_ANSIBLE_audit_rules_unsuccessful_file_modification_o_trunc_write]: @jan-cerny fixed

File: shared/templates/create_audit_rules_unsuccessful_file_modification_detailed.py (hunk 1 above)
Comment (redhatrises) [reply shared/templates/create_audit_rules_unsuccessful_file_modification_detailed.py]: @jan-cerny done

**Comment by jan-cerny** [comment]
Regarding the bash remediations, for RHEL and OL this PR has been replaced by https://github.com/ComplianceAsCode/content/pull/4107/files. 

What about other platforms? Wouldn't it be better to just add platforms to list of platforms to remediation scripts introduced in #4107 ?

Should we use the Ansible template introduced here to have Ansible remediations for those rules?

**Comment by jan-cerny** [comment]
Any opinions?

**Comment by jan-cerny** [comment]
bump @redhatrises

**Comment by redhatrises** [comment]
Not sure what the ask is here. I consider #4107 as an interim fix, and the fixes should be updated to not rely on having `ospp.rules` installed at all.

**Comment by jan-cerny** [comment]
AFAIK using `fix_audit_syscall_rule` doesn't ensure ordering of audit rules. The patch in #4107 ensures the order, so this change would be a regression.

I see #4107 was introduced as a "provisional" fix. @matejak or @yuumasato , do you remember what is a proper fix for audit_rules_unsuccessful_file_modification_detailed?

>  and the fixes should be updated to not rely on having ospp.rules installed at all.

We don't rely on the file installed, we have included the contents of the file in the remediation. https://github.com/ComplianceAsCode/content/blob/master/shared/bash_remediation_functions/create_audit_remediation_unsuccessful_file_modification_detailed.sh

**Comment by yuumasato** [comment]
> I see #4107 was introduced as a "provisional" fix. @matejak or @yuumasato , do you remember what is a proper fix for audit_rules_unsuccessful_file_modification_detailed?

Well...The proper fix is to have Bash and Ansible scripts that can re-order the audit rules if all of them are present. By all of them, consider all of the audit rules specific for a syscall.

**Comment by yuumasato** [comment]
I'm removing Blocker label from this PR, as we have an interim fix #4107.

**Comment by jan-cerny** [comment]
All the affected rules already have remediations, and this one doesn't address the problem of ordering rules anyway,  so I suggest closing it.

```
[jcerny@thinkpad scap-security-guide{master}]$ find . -name audit_rules_unsuccessful_file_modification*_o_creat | xargs tree 
./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_open_by_handle_at_o_creat
├── ansible
│   └── shared.yml
├── bash
│   └── shared.sh
└── rule.yml
./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_open_o_creat
├── ansible
│   └── shared.yml
├── bash
│   └── shared.sh
└── rule.yml
./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_openat_o_creat
├── ansible
│   └── shared.yml
├── bash
│   └── shared.sh
└── rule.yml

6 directories, 9 files
[jcerny@thinkpad scap-security-guide{master}]$ find . -name audit_rules_unsuccessful_file_modification*_o_trunc_write | xargs tree 
./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_open_o_trunc_write
├── ansible
│   └── shared.yml
├── bash
│   └── shared.sh
└── rule.yml
./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_open_by_handle_at_o_trunc_write
├── ansible
│   └── shared.yml
├── bash
│   └── shared.sh
└── rule.yml
./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_openat_o_trunc_write
├── ansible
│   └── shared.yml
├── bash
│   └── shared.sh
└── rule.yml

6 directories, 9 files
```

**Comment by redhatrises** [comment]
> All the affected rules already have remediations, and this one doesn't address the problem of ordering rules anyway, so I suggest closing it.
> 
> ```
> [jcerny@thinkpad scap-security-guide{master}]$ find . -name audit_rules_unsuccessful_file_modification*_o_creat | xargs tree 
> ./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_open_by_handle_at_o_creat
> ├── ansible
> │   └── shared.yml
> ├── bash
> │   └── shared.sh
> └── rule.yml
> ./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_open_o_creat
> ├── ansible
> │   └── shared.yml
> ├── bash
> │   └── shared.sh
> └── rule.yml
> ./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_openat_o_creat
> ├── ansible
> │   └── shared.yml
> ├── bash
> │   └── shared.sh
> └── rule.yml
> 
> 6 directories, 9 files
> [jcerny@thinkpad scap-security-guide{master}]$ find . -name audit_rules_unsuccessful_file_modification*_o_trunc_write | xargs tree 
> ./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_open_o_trunc_write
> ├── ansible
> │   └── shared.yml
> ├── bash
> │   └── shared.sh
> └── rule.yml
> ./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_open_by_handle_at_o_trunc_write
> ├── ansible
> │   └── shared.yml
> ├── bash
> │   └── shared.sh
> └── rule.yml
> ./linux_os/guide/system/auditing/auditd_configure_rules/audit_file_modification/audit_rules_unsuccessful_file_modification_openat_o_trunc_write
> ├── ansible
> │   └── shared.yml
> ├── bash
> │   └── shared.sh
> └── rule.yml
> 
> 6 directories, 9 files
> ```

Current rules are interim, copy a config file, and need to be replaced without requiring a config file. Ordering happens through a different rule.

**Comment by jan-cerny** [comment]
> Current rules are interim, copy a config file, and need to be replaced without requiring a config file. 

Yes, I know, I also agree that copying a file is wrong approach.

> Ordering happens through a different rule.

The rules `audit_rules_unsuccessful_file_modification_openat_rule_order`, `audit_rules_unsuccessful_file_modification_open_by_handle_at_rule_order`,
`audit_rules_unsuccessful_file_modification_open_rule_order` have only bash remediation which again copy the config file. 
My understanding is that this remediation would overwrite the effect of the remediation you propose.

@yuumasato said that:

> The proper fix is to have Bash and Ansible scripts that can re-order the audit rules if all of them are present. By all of them, consider all of the audit rules specific for a syscall.

@redhatrises It isn't clear to me how this PR improves the situation. How does it ensure that the remediation puts the audit rule in the right place in the audit config file?

**Comment by adelton** [comment]
Could you please rebase on the latest master?

**Comment by redhatrises** [comment]
Closing this PR for now. This is going to have to be done anyways for all audit rules.
